src/libraries/System.Linq/src/System/Linq/Distinct.cs

```diff
@@ -20,6 +20,34 @@ public static IEnumerable<TSource> Distinct<TSource>(this IEnumerable<TSource> s
             return new DistinctIterator<TSource>(source, comparer);
         }
 
+        public static IEnumerable<TSource> DistinctBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector) => DistinctBy(source, keySelector, null);
+
+        public static IEnumerable<TSource> DistinctBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? comparer)
+        {
+            if (source is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+            if (keySelector is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.keySelector);
+            }
+
+            return DistinctByIterator(source, keySelector, comparer);
+        }
+
+        private static IEnumerable<TSource> DistinctByIterator<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? comparer)
+        {
+            var set = new HashSet<TKey>(DefaultInternalSetCapacity, comparer);
+            foreach (TSource element in source)
+            {
+                if (set.Add(keySelector(element)))
+                {
+                    yield return element;
+                }
+            }
```

Comment:
Not a big deal, but it'd be nice to avoid the HashSet allocation if the source ends up being empty:
```C#
IEnumerator<TSource> e = source.GetEnumerator();
try
{
    if (e.MoveNext())
    {
        do
        {
            TSource current = e.Current;
            if (see.Add(keySelector(current)))
            {
                yield return current;
            }
        }
        while (e.MoveNext());
    }
}
finally
{
    e.Dispose();
}

Reply:
Done. I didn't apply similar changes to `IntersectBy`/`ExceptBy`/`UnionBy` since it would change the evaluation semantics of the `second` enumerable.